README
NAME
    CGI::Emulate::PSGI - PSGI adapter for CGI

SYNOPSIS
        my $app = CGI::Emulate::PSGI->handler(sub {
            # Existing CGI code
        });

DESCRIPTION
    This module allows an application designed for the CGI environment to
    run in a PSGI environment, and thus on any of the backends that PSGI
    supports.

    It works by translating the environment provided by the PSGI
    specification to one expected by the CGI specification. Likewise, it
    captures output as it would be prepared for the CGI standard, and
    translates it to the format expected for the PSGI standard using
    CGI::Parse::PSGI module.

CGI.pm
    If your application uses CGI, be sure to cleanup the global variables in
    the handler loop yourself, so:

        my $app = CGI::Emulate::PSGI->handler(sub {
            use CGI;
            CGI::initialize_globals();
            my $q = CGI->new;
            # ...
        });

    Otherwise previous request variables will be reused in the new requests.

    Alternatively, you can install and use CGI::Compile from CPAN and
    compiles your existing CGI scripts into a sub that is perfectly ready to
    be converted to PSGI application using this module.

      my $sub = CGI::Compile->compile("/path/to/script.cgi");
      my $app = CGI::Emulate::PSGI->handler($sub);

    This will take care of assigning an unique namespace for each script
    etc. See CGI::Compile for details.

    You can also consider using CGI::PSGI but that would require you to
    slightly change your code from:

      my $q = CGI->new;
      # ...
      print $q->header, $output;

    into:

      use CGI::PSGI;

      my $app = sub {
          my $env = shift;
          my $q = CGI::PSGI->new($env);
          # ...
          return [ $q->psgi_header, [ $output ] ];
      };

    See CGI::PSGI for details.

METHODS
    handler
          my $app = CGI::Emulate::PSGI->handler($code);

        Creates a PSGI application code reference out of CGI code reference.

    emulate_environment
          my %env = CGI::Emulate::PSGI->emulate_environment($env);

        Creates an environment hash out of PSGI environment hash. If your
        code or framework just needs an environment variable emulation, use
        this method like:

          local %ENV = (%ENV, CGI::Emulate::PSGI->emulate_environment($env));
          # run your application

        If you use "handler" method to create a PSGI environment hash, this
        is automatically called in the created application.

AUTHOR
    Tokuhiro Matsuno <tokuhirom@cpan.org>

    Tatsuhiko Miyagawa

COPYRIGHT AND LICENSE
    Copyright (c) 2009-2010 by tokuhirom.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.

SEE ALSO
    PSGI CGI::Compile CGI::PSGI Plack CGI::Parse::PSGI



Changes
{{$NEXT}}

0.23  2017-05-08 04:45:43 UTC

        - Respect an NPH script's HTTP status line (wu-lee) #4

0.22  2016-07-19 08:03:23 PDT
        - Do not set HTTP_PROXY (kazeburo) #14

0.21  2015-07-21 09:02:04 PDT
        - Fix to restore the selected filehandle correctly #12

0.20  2014-09-08 11:06:19 PDT
        - Add Test::Requires dependency (andk)

0.19  2014-08-31 14:04:44 JST
        - Revert #3 since it causes deep recursion when printing to STDERR
          inside CGI handler under some perl environments. #6

0.18  2014-08-07 07:58:32 PDT
        - Fix perl prereq (haarg) #7

0.17  2014-08-03 18:30:54 PDT
        - Make it work with debugger with perl 5.8.x (tehmoth) #5

0.16  2013-09-18 12:23:21 JST
        - Use tied handle for STDIN/STDERR (#3)
        - Convert to Milla

0.15 Fri Oct 19 15:09:17 PDT 2012
     - Repackage with the latest Module::Install

0.14 Mon Oct 31 10:37:49 PDT 2011
     - Fixed a bug where croak was not imported from Carp (ctfliblime)

0.13 Sun Sep 18 12:10:56 PDT 2011
     - Set the status header to 302 if there's Location header but no Status in the resonse (ASP)

0.12 Thu Jun  9 23:58:10 PDT 2011
     - Fixed the PSGI header generation to prevent invalid PSGI response headers
       such as newlines in the value and "Status" key

0.11 Fri Feb 18 21:35:29 PST 2011
     - Filter psgix.* environment too (mkanat)

0.10 Sun May 16 00:39:55 PDT 2010
     - Added a possible workaround to deal with Win32 systems to undo :utf8 PerlIO layer
       perl #75106 (suggested by Eric Brine)

0.09 Sat May  8 18:52:04 PDT 2010
     - Added a test for utf8 decoding

0.08 Sat May  1 04:30:25 PDT 2010
     - Fixed Test::More deps

0.07 Tue Apr 27 15:44:08 PDT 2010
     - binmode the stdout filehandle when parsing

0.06 Fri Mar  5 18:38:45 PST 2010
     - Selects main::STDOUT before running the CGI handler to avoid some gotchas when
       other file handles such as the temporary file handles are selected. (reported by athomason)

0.05 Wed Jan  6 21:03:15 PST 2010
     - Factored out CGI output parser into CGI::Parse::PSGI as well as
       environment emulation as ->emulate_environment method (kazuho)

0.04 Fri Dec 11 22:59:14 PST 2009
     - Do not override existing %ENV values (kazuho)

0.03 Sun Dec  6 06:05:25 GMT 2009
     - Added docs about CGI::Compile
     - Fixed a warning for SEEK_SET imported twice

0.02
    - fixed packaging issue

0.01

    - initial release



lib/CGI/Parse/PSGI.pm
package CGI::Parse::PSGI;
use strict;
use base qw(Exporter);
our @EXPORT_OK = qw( parse_cgi_output );

use IO::File; # perl bug: should be loaded to call ->getline etc. on filehandle/PerlIO
use HTTP::Response;

our %DEFAULT_OPTS = (
    ignore_status_line => 0,
);

sub parse_cgi_output {
    my $output = shift;
    my $options = \%DEFAULT_OPTS;
    if (ref $_[0] eq 'HASH') {
        $options = { %DEFAULT_OPTS, %{ +shift } }; # Use default opts where none supplied
    }

    my $length;
    if (ref $output eq 'SCALAR') {
        $length = length $$output;
        open my $io, "<", $output;
        $output = $io;
    } else {
        open my $tmp, '<&=:perlio:raw', fileno($output) or die $!;
        $output = $tmp;
        $length = -s $output;
    }

    my $headers;
    while ( my $line = $output->getline ) {
        $headers .= $line;
        last if $headers =~ /\x0d?\x0a\x0d?\x0a$/;
    }
    unless ( defined $headers ) {
        $headers = "HTTP/1.1 500 Internal Server Error\x0d\x0a";
    }

    unless ( $headers =~ /^HTTP/ ) {
        $headers = "HTTP/1.1 200 OK\x0d\x0a" . $headers;
    }

    my $response = HTTP::Response->parse($headers);

    # RFC 3875 6.2.3
    if ($response->header('Location') && !$response->header('Status')) {
        $response->header('Status', 302);
    }

    my $status = $options->{ignore_status_line}?
        200 : ($response->code || 200);

    my $status_header = $response->header('Status');
    if ($status_header) {
        # Use the header status preferentially, if present and well formed

        # Extract the code from the header (should be 3 digits, non zero)
        my ($code) = ($status_header =~ /^ \s* (\d+) /x);

        $status = $code || $status;
    }

    $response->remove_header('Status'); # PSGI doesn't allow having Status header in the response

    my $remaining = $length - tell( $output );
    if ( $response->code == 500 && !$remaining ) {
        return [
            500,
            [ 'Content-Type' => 'text/html' ],
            [ $response->error_as_HTML ]
        ];
    }

    $response->content_length($remaining);

    return [
        $status,
        +[
            map {
                my $k = $_;
                map { ( $k => _cleanup_newline($_) ) } $response->headers->header($_);
            } $response->headers->header_field_names
        ],
        $output,
    ];
}

sub _cleanup_newline {
    local $_ = shift;
    s/\r?\n//g;
    return $_;
}

1;

__END__

=head1 NAME

CGI::Parse::PSGI - Parses CGI output and creates PSGI response out of it

=head1 DESCRIPTION

  use CGI::Parse::PSGI qw(parse_cgi_output);

  my $output = YourApp->run;
  my $psgi_res = parse_cgi_output(\$output);

An option hash can also be passed:

  my $psgi_res = parse_cgi_output(\$output, \%options);

=head1 SYNOPSIS

CGI::Parse::PSGI exports one function C<parse_cgi_output> that takes a
filehandle or a reference to a string to read a CGI script output, and
creates a PSGI response (an array reference containing status code,
headers and a body) by reading the output.

Use L<CGI::Emulate::PSGI> if you have a CGI I<code> not the I<output>,
which takes care of automatically parsing the output, using this
module, from your callback code.

=head1 OPTIONS

As mentioned above, C<parse_cgi_output> can accept an options hash as
the second argument.

Currently the options available are:

=over 4

=item C<ignore_status_line>

A boolean value, defaulting to 0 (false). If true, the status in the
HTTP protocol line is not used to set the default status in absence of
a status header.

=back

The options can be supplied to earlier versions, and will be ignored
without error.  Hence you can preserve legacy behaviour like this:

    parse_cgi_output(\$output, {ignore_status_line => 1});

This will ensure that if the script output includes an edge case
like this:

    HTTP/1.1 666 SNAFU
    Content-Type: text/plain

    This should be OK!

then the old behaviour of ignoring the status line and returning 200
is preserved.

=head1 AUTHOR

Tatsuhiko Miyagawa

=head1 SEE ALSO

L<CGI::Emulate::PSGI>

=cut


lib/CGI/Emulate/PSGI/Streaming.pm
package CGI::Emulate::PSGI::Streaming;
use strict;
use warnings;
use parent 'CGI::Emulate::PSGI';
use CGI::Parse::PSGI;
use SelectSaver;
use Carp qw(croak);
use 5.008001;

our $VERSION = '0.21';

sub handler {
    my ($class, $code, ) = @_;

    return sub {
        my $env = shift;

        my $wrapped_app = sub {
            my ($stdout) = @_;
            my $saver = SelectSaver->new("::STDOUT");
            local %ENV = (%ENV, $class->emulate_environment($env));

            local *STDIN  = $env->{'psgi.input'};
            local *STDOUT = $stdout;
            local *STDERR = $env->{'psgi.errors'};

            $code->();
            close $stdout;
        };

        return sub {
            my ($responder) = @_;

            CGI::Parse::PSGI::parse_cgi_output_streaming($responder,$wrapped_app);
        };
    };
}

1;


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.006.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "PSGI adapter for CGI",
  "AUTHOR" => "Tokuhiro Matsuno <tokuhirom\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "CGI-Emulate-PSGI",
  "LICENSE" => "perl",
  "MIN_PERL_VERSION" => "5.008001",
  "NAME" => "CGI::Emulate::PSGI",
  "PREREQ_PM" => {
    "CGI" => "3.63",
    "HTTP::Response" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.88",
    "Test::Requires" => "0.08"
  },
  "VERSION" => "0.22",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "CGI" => "3.63",
  "HTTP::Response" => 0,
  "Test::More" => "0.88",
  "Test::Requires" => "0.08"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
